UChicago VuFind.js

```diff
@@ -0,0 +1,794 @@
+		// if there's a 502 field, it should be a thesis
+		const isDissertation = marc => field_exists('502')(marc);
+		// if the record type is 'p', it's a manuscript
+		const isManuscript = marc => marc.substring(6,7) == 'p';
+		// if the item type is film, it's a film
+		const isFilm = item => item.itemType === 'film';
+		// if the record type is 'e', it's a map
+		const isMap = marc => marc.substring(6,7) == 'e';
+
+		if (isDissertation(marc)) {
+		// a dissertation is a 'thesis'
+		item.itemType = "thesis";
+		} else if (isManuscript(marc)) {
+		// a manuscript is a 'manuscript';
+		item.itemType = "manuscript";
+		} else if (isFilm(item)) {
+		// a film is a 'videoRecording' since we don't have
+		// any film prints
+		item.itemType = "videoRecording";
+		} else if (isMap(marc)) {
+		// a map is a 'map';
+		item.itemType = "map";
+		}
+	}
+
+	// perform the fixes
+	addUrl(item);
+	fixItemType(item);
+	updateCN(item);
+	
+	}
+	
+	// this part is based on the Finna translator code
+	let cleanURL = url.replace(/[#?].*$/, '').replace(/\/$/, '');
+	let marcURL = cleanURL + '/Export?style=MARC';
+	
+	// use MARC import translator to ingest binary MARC records
+	ZU.doGet(marcURL, function (marcData) {
+	var success = false;
+	var translator = Zotero.loadTranslator("import");
+	translator.setTranslator("a6ee60df-1ddc-4aae-bb25-45e0537be973");
+	translator.setString(marcData);
+	translator.setHandler('itemDone', function (_, item) {
+		if (item.place) {
+		item.place = item.place.replace(/\[[^[]+\]/, '');
+		}
+		
+		if (item.publisher) {
+		item.publisher = item.publisher.replace(/&amp;/g, '&');
+		}
+		success = true;
+
+		// apply the above UChicago customizations
+		customizeMARC(doc)(item)(marcData);
+		item.complete();
+
+	});
+
+	translator.setHandler('done', () => {});
+	translator.translate();
+	});
+}
+
+const getSearchResults = doc => {
+	// get every search result DOM element
+	const rowNodes = doc.querySelectorAll('li[id^=result]');
+	// make the node list into an array
+	const a = Array.from(rowNodes);
+	let obj = {};
+	// extract information from each li element for output
+	const buildOutput = r => {
+	const linkElement = r.querySelector('.title.getFull');
+	const entry_url = linkElement.href;
+	const title = ZU.trimInternal(linkElement.textContent);
+	if (entry_url && title) { 
+		obj[entry_url] = title;
+	}
+	}
+	a.map(buildOutput);
+	return obj;
+}
+
+const detectWeb = (doc, url) => {
+	// VuFind URL patterns starting with 'Record' are for single items
+	if (url.includes('Record')) {
+	if (doc.querySelector('.format.video')) {
+		return 'videoRecording';
+	} else if (doc.querySelector('.format.dissertations')) {
+		return 'thesis';
+	} else if (doc.querySelector('.format.audio')) {
+		return 'audioRecording';
+	} else if (doc.querySelector('.format.map')) {
+		return 'map';
+	} else {
+		return 'book';
+	}
+	// VuFind URL patterns starting with 'Search' are for search results
+	} else if (url.includes('Search')) {
```

Comment:
Can we make these a little more specific? Like, if we're trying to match a URL that looks like `https://example.com/search?q=123`, `url.includes('/search?')` is a lot better and less likely to hit false positives than `url.includes('search')`.